flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java

```diff
@@ -307,14 +340,23 @@ public void abortCheckpoint(long abortedCheckpointId) {
                 jobVertexID,
                 subtaskIndex);
 
+        // delete the referenced checkpoint meantime
         pruneCheckpoints(
-                snapshotCheckpointId -> snapshotCheckpointId == abortedCheckpointId, false);
+                snapshotCheckpointId ->
+                        (snapshotCheckpointId == abortedCheckpointId
+                                || referredByCheckpointID.getOrDefault(snapshotCheckpointId, -1L)
+                                        == abortedCheckpointId),
```

Comment:
Do I understand correctly that in case of changelog
- discard does nothing for jm part, because of the empty 
`discardState()` in `ChangelogStateBackendHandleImpl`
- the local part can is discarded, although it could be re-used; but it won't hurt correctness (on recovery, non-local state will be used)

?

Reply:
> 1. discard does nothing for jm part, because of the empty
discardState() in ChangelogStateBackendHandleImpl

Partially agree：
- For local state: discard does nothing for jm part, because of the empty
`discardState()` in `ChangelogStateBackendHandleImpl` and **no local handle would be registered in** `registerSharedStates(SharedStateRegistry stateRegistry, long checkpointID)`
- For remote state: `SharedStateRegistry` on JM would discard changelogStateHandle.

> 2. the local part can is discarded, although it could be re-used; but it won't hurt correctness (on recovery, non-local state will be used)

Yes, local recovery is nice to have but not a must. `BackendRestorerProcedure#createAndRestore()` would try to restore from remote.